### PR TITLE
feat: add cp transfer progress reporting

### LINKED ISSUE #168: feat: progress indicator for cp file transfers
## Summary

`ssmctl cp` gives no feedback during a transfer. For the in-band chunked path (multiple SSM `SendCommand` calls) and especially for the S3-backed path, there is no indication of progress and the command can appear to hang for large files.

## Proposed Behaviour

**In-band (chunked) transfers:**
Display a progress bar or byte counter that updates as each chunk is sent/received:
```
Uploading web-1:/var/app/deploy.tar.gz ... 1.2 MB / 2.0 MB [=======>    ] 60%
```
The total size is known upfront (`os.ReadFile` / `os.Stat`), and each chunk is a fixed 3072 bytes of base64 (representing ~2304 bytes of raw data), so progress can be computed exactly.

**S3-backed transfers:**
Display a spinner with a running bytes count whilst the S3 `PutObject` / `GetObject` streams:
```
Uploading to S3 staging ... 8.4 MB / 45.2 MB [=====>      ] 18%
```
The S3 SDK supports `Body` as an `io.Reader`, so wrapping it in a `progressReader` gives accurate streaming progress.

**Non-TTY / JSON output:**
Progress output should be suppressed when stdout is not a TTY or when `--output json` is set, so that piped usage and CI are unaffected.

## Implementation Notes

- Consider using [`github.com/schollz/progressbar/v3`](https://github.com/schollz/progressbar) — zero heavy dependencies, TTY-aware, works on all platforms
- Alternatively a simple hand-rolled `\r`-based counter avoids a new dependency
- Progress should write to stderr so it does not interfere with `--output json` to stdout

### PR DESCRIPTION
**Pull request by pollychen-lab**
## Summary

- Add progress reporting hooks for cp transfers without changing existing transfer APIs.
- Render byte-count progress to stderr for interactive text output only.
- Suppress progress for JSON output and non-TTY stdout so pipes and automation stay quiet.
- Cover in-band upload, S3 upload/download progress, and command-layer formatting/suppression.

Fixes #168.

## Validation

- `docker run --rm -u $(id -u):$(id -g) -e GOCACHE=/tmp/go-build -e GOPATH=/tmp/go -v "$PWD":/workspace -w /workspace golang:1.26.4 go test ./internal/ssm ./internal/cmd`
- `docker run --rm -u $(id -u):$(id -g) -e GOCACHE=/tmp/go-build -e GOPATH=/tmp/go -v "$PWD":/workspace -w /workspace golang:1.26.4 go test ./e2e/ -v -count=1`
- `docker run --rm -u $(id -u):$(id -g) -e GOCACHE=/tmp/go-build -e GOPATH=/tmp/go -v "$PWD":/workspace -w /workspace golang:1.26.4 go vet ./...`
- `docker run --rm -u $(id -u):$(id -g) -e GOCACHE=/tmp/go-build -e GOPATH=/tmp/go -v "$PWD":/workspace -w /workspace golang:1.26.4 go test ./...`
- `docker run --rm -u $(id -u):$(id -g) -e GOCACHE=/tmp/go-build -e GOPATH=/tmp/go -v "$PWD":/workspace -w /workspace golang:1.26.4 go build -o bin/ssmctl ./cmd/ssmctl`
- `docker run --rm -u $(id -u):$(id -g) -e GOCACHE=/tmp/go-build -e GOPATH=/tmp/go -v "$PWD":/workspace -w /workspace golang:1.26.4 go test -race -coverprofile=coverage.out ./cmd/... ./internal/...`

### CODE REVIEW

File: internal/ssm/progress.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	progress ProgressFunc
+}
+
+func (r *progressReader) Read(p []byte) (int, error) {
+	n, err := r.reader.Read(p)
+	if n > 0 {
+		r.done += int64(n)
+		reportProgress(r.progress, r.done, r.total)
+	}
+	return n, err
```
Comment (rhysmcneill) [inline internal/ssm/progress.go]: Hey @pollychen-lab, can you check this. After all CI's are green I will conduct a thorough review.

**Review by rhysmcneill — CHANGES_REQUESTED** [review]

File: internal/ssm/progress.go (hunk 1 above)
Comment (pollychen-lab) [reply internal/ssm/progress.go]: thank you for reviewing @rhysmcneill

File: internal/cmd/cp_progress.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	}
+}
+
+func isTerminal(w io.Writer) bool {
+	file, ok := w.(*os.File)
+	if !ok {
+		return false
+	}
+	stat, err := file.Stat()
+	return err == nil && stat.Mode()&os.ModeCharDevice != 0
```
Comment (rhysmcneill) [inline internal/cmd/cp_progress.go:60]: Nice TTY suppression logic overall, but `os.ModeCharDevice` matches **any** character device (including `/dev/null`, `/dev/random`, `/dev/zero`), not strictly a terminal. The test below reinforces this by asserting that `/dev/null` "is a terminal", which is semantically wrong — it's a character device, not a TTY.

Rather than adding an external dependency to fix this properly, the simplest safe change is to rename `isTerminal` → `isCharDevice` and update the test assertion text to match (`"did not report as a character device"` is actually already the right words — it's the test failure message that reads like it expects a terminal). A short comment on the function explaining the heuristic would also help future readers.

```go
// isCharDevice reports whether w is a character device.
// We use this as a TTY heuristic: in practice cmd.OutOrStdout() is
// os.Stdout (a character device) when running interactively.
func isCharDevice(w io.Writer) bool {
```

File: internal/cmd/cp_progress.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+	return newProgressWriter(cmd.ErrOrStderr(), action)
+}
+
+func newProgressWriter(w io.Writer, action string) ssmlib.ProgressFunc {
+	var lastDone, lastTotal int64 = -1, -2
+	var finished bool
+
+	return func(done, total int64) {
+		if finished || done == lastDone && total == lastTotal {
```
Comment (rhysmcneill) [inline internal/cmd/cp_progress.go]: Nit: Go evaluates `&&` before `||` so this is already correct, but explicit parentheses make the intent immediately obvious to a reader without needing to recall precedence rules:

```go
if finished || (done == lastDone && total == lastTotal) {
```

File: internal/cmd/cp_progress.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+			return
+		}
+		lastDone, lastTotal = done, total
+
+		if total >= 0 {
+			percent := int64(100)
+			if total > 0 {
```
Comment (rhysmcneill) [inline internal/cmd/cp_progress.go]: Edge case: when `total = 0` (empty file), `percent` keeps its initial value of `100` because the `total > 0` guard skips the calculation. This means a zero-byte transfer would show `0 B / 0 B (100%)` on the very first event.

Changing the guard to `if done > 0 && total > 0` would avoid the false 100% and just display `0 B / 0 B (0%)` instead, which is more accurate.

File: internal/ssm/transfer.go [hunk 5]
```diff
@@ -121,6 +129,11 @@ func Upload(ctx context.Context, client RunAPI, target TargetInfo, localPath, re
 			return nil, fmt.Errorf("chunk command failed: %s", result.Stderr)
 		}
 		chunks++
+		rawDone := int64(end / 4 * 3)
+		if rawDone > totalBytes {
```
Comment (rhysmcneill) [inline internal/ssm/transfer.go:133]: Progress reaches `rawDone = totalBytes` (100%) at the end of the final chunk here, but the `finalise` step — the `mv` / `Move-Item` that atomically moves the temp file to its destination — runs after this with no feedback. For large files on slower remote filesystems there will be a noticeable pause showing 100% before the command actually returns.

Consider either holding back the final `reportProgress` call until after `finaliseCmd` succeeds, or emitting a brief "Finalising..." status during the move.

File: internal/ssm/transfer_test.go [hunk 6]
```diff
@@ -135,6 +135,42 @@ func TestUpload(t *testing.T) {
...
+	if len(events) < 3 {
+		t.Fatalf("progress events = %v, want initial, chunk, and completion events", events)
+	}
+	if events[0] != [2]int64{0, int64(len(content))} {
+		t.Fatalf("first progress event = %v, want 0/%d", events[0], len(content))
+	}
+	if got := events[len(events)-1]; got != [2]int64{int64(len(content)), int64(len(content))} {
+		t.Fatalf("final progress event = %v, want %d/%d", got, len(content), len(content))
+	}
+}
```
Comment (rhysmcneill) [inline internal/ssm/transfer_test.go:204]: `DownloadWithOptions` fires a single terminal progress event at the end of `DownloadWithOptions` (`reportProgress(opts.Progress, int64(len(data)), int64(len(data)))`), but there's no test covering that path — `TestUploadWithOptionsReportsProgress` above has an equivalent, so a matching `TestDownloadWithOptionsReportsProgress` would be a natural pair here. This is also very likely where some of the 12 lines flagged by Codecov are coming from.

**Review by rhysmcneill — CHANGES_REQUESTED** [review]
Thanks @pollychen-lab — this is a really clean implementation. The `TransferOptions` / `ProgressFunc` callback design keeps the `internal/ssm` layer completely free of UI concerns, the backward-compatible `WithOptions` wrapper pattern is exactly right, and the test coverage is solid. A few things to address before merge.

**Comment by pollychen-lab** [comment]
> Thanks @pollychen-lab — this is a really clean implementation. The `TransferOptions` / `ProgressFunc` callback design keeps the `internal/ssm` layer completely free of UI concerns, the backward-compatible `WithOptions` wrapper pattern is exactly right, and the test coverage is solid. A few things to address before merge.

on it

**Review by rhysmcneill — APPROVED** [review]
Thanks for the swift turnaround on all the feedback @pollychen-lab - every point from the previous review has been addressed cleanly:

- `isTerminal` renamed to `isCharDevice` with a comment explaining the heuristic - no new dependency needed.
- Explicit parentheses on the dedup guard.
- Zero-byte edge case fixed (`percent` defaults to `0`, guarded by `done > 0 && total > 0`) and covered by the new `TestProgressWriterFormatsZeroByteTransfer` test.
- Upload progress now correctly fires 100% only after `finaliseCmd` succeeds. The intermediate loop suppresses the final report (`rawDone < totalBytes`) and the post-finalise `reportProgress` handles it. The updated `TestUploadWithOptionsReportsProgress` asserting `commandsSeen` at the time of the final event is a great addition.
- `TestDownloadWithOptionsReportsProgress` fills the coverage gap.

Happy with this - approving.

**Comment by pollychen-lab** [comment]
@rhysmcneill thank you so much
